README.md
# ocb

`ocb` (OCaml Badgen) is an [OCaml] library for [SVG] badge generation. There's also a command-line tool provided.

## Quick start

### Using the command-line tool

```sh
ocb --label Hello --color green --style flat --labelcolor white --status Goodbye > out.svg && xdg-open out.svg
```

Gives the following result: [![cli example](example/cli.svg)](example/cli.svg).

### Using the library

```ml
let () =
  let open Ocb in
  Gen.mk Format.std_formatter ~label:"Hello" ~color:Color.Blue ~style:Style.Flat ~label_color:Color.White ~status:"Goodbye" ()
```

## Credits

This project is inspired by [badgen].

[badgen]: https://github.com/badgen/badgen
[OCaml]: https://ocaml.org/
[SVG]: https://en.wikipedia.org/wiki/Scalable_Vector_Graphics


CHANGES.md
## 0.2 - 2022-06-21

- fix unerasable-optional-argument warnings for the `Gen.bare` function
- add a unique id to avoid rendering issues
- change the order of the svg attributes to make some linters happy

## 0.1 - 2021-01-07

- first release


src/gen.ml
let counter = ref 0

let calc_width =
  let table = Verdana.table in
  let len = Array.length table in
  let fallback = table.(64) in
  fun text ->
    let total = ref 0. in
    String.iter
      (fun c ->
        let code = Char.code c in
        let width = if code >= len then fallback else table.(code) in
        total := !total +. width )
      text;
    !total

let sanitize s =
  let buffer = Buffer.create @@ String.length s in
  let fmt = Format.formatter_of_buffer buffer in
  let add = Format.fprintf fmt in
  let add_c = Buffer.add_char buffer in
  String.iter
    (function
      | '&' -> add "&amp;"
      | '<' -> add "&lt;"
      | '>' -> add "&gt;"
      | '\'' -> add "&apos;"
      | '\"' -> add "&quot;"
      | '@' -> add "&commat;"
      | c -> add_c c )
    s;
  Buffer.contents buffer

let create_accessible_text label status =
  if String.equal label "" then status else Format.sprintf "%s: %s" label status

let bare fmt ?(color = Color.Blue) ?(style = Style.Classic) ?(scale = 1.)
    ~status () =
  let st_text_width = calc_width status in
  let st_rect_width = st_text_width +. 115. in
  let status = sanitize status in
  let width = scale *. st_rect_width /. 10. in
  let height = scale *. 20. in
  match style with
  | Classic ->
    Format.fprintf fmt
      {|<svg xmlns="http://www.w3.org/2000/svg" width="%f" height="%f" viewBox="0 0 %f 200" role="img" aria-label="%s">
  <title>%s</title>
  <linearGradient id="a" x2="0" y2="100%%">
    <stop offset="0" stop-opacity=".1" stop-color="#EEE"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <mask id="m"><rect width="%f" height="200" rx="30" fill="#FFF"/></mask>
  <g mask="url(#m)">
    <rect width="%f" height="200" fill="#%a" x="0"/>
    <rect width="%f" height="200" fill="url(#a)"/>
  </g>
  <g aria-hidden="true" fill="#fff" text-anchor="start" font-family="Verdana,DejaVu Sans,sans-serif" font-size="110">
    <text x="65" y="148" textLength="%f" fill="#000" opacity="0.25">%s</text>
    <text x="55" y="138" textLength="%f">%s</text>
  </g>
</svg>@.|}
      width height st_rect_width status status st_rect_width st_rect_width
      Color.pp color st_rect_width st_text_width status st_text_width status
  | Flat ->
    Format.fprintf fmt
      {|<svg xmlns="http://www.w3.org/2000/svg" width="%f" height="%f" viewBox="0 0 %f 200" role="img" aria-label="%s">
  <title>%s</title>
  <g>
    <rect fill="#%a" x="0" width="%f" height="200"/>
  </g>
  <g aria-hidden="true" fill="#fff" text-anchor="start" font-family="Verdana,DejaVu Sans,sans-serif" font-size="110">
    <text x="65" y="148" textLength="%f" fill="#000" opacity="0.1">%s</text>
    <text x="55" y="138" textLength="%f">%s</text>
  </g>
</svg>@.|}
      width height st_rect_width status status Color.pp color st_rect_width
      st_text_width status st_text_width status

let mk fmt ?(counter = counter) ?(label = "") ?(color = Color.Blue)
    ?(style = Style.Classic) ?(icon = None) ?(icon_width = 13.)
    ?(label_color = Color.Custom "555") ?(scale = 1.) ~status () =
  if String.equal label "" && Option.is_none icon then
    bare fmt ~status ~color ~style ~scale ()
  else
    let icon_width = icon_width *. 10. in
    let icon_span_width =
      if Option.is_none icon then 0.
      else if String.equal label "" then icon_width -. 18.
      else icon_width +. 30.
    in
    let sb_text_start =
      if Option.is_none icon then 50. else icon_width +. 50.
    in
    let sb_text_width = calc_width label in
    let st_text_width = calc_width status in
    let sb_rect_width = sb_text_width +. 100. +. icon_span_width in
    let st_rect_width = st_text_width +. 100. in
    let width = sb_rect_width +. st_rect_width in
    let xlink =
      if Option.is_none icon then ""
      else {| xmlns:xlink="http://www.w3.org/1999/xlink"|}
    in
    let label = sanitize label in
    let status = sanitize status in
    let accessible_text = create_accessible_text label status in
    let svg_width = scale *. width /. 10. in
    let svg_height = scale *. 20. in
    match style with
    | Classic ->
      incr counter;
      Format.fprintf fmt
        {|<svg xmlns="http://www.w3.org/2000/svg"%s width="%f" height="%f" viewBox="0 0 %f 200" role="img" aria-label="%s">
  <title>%s</title>
  <linearGradient id="ocaml-ocb-a-%d" x2="0" y2="100%%">
    <stop offset="0" stop-opacity=".1" stop-color="#EEE"/>
    <stop offset="1" stop-opacity=".1"/>
  </linearGradient>
  <mask id="ocaml-ocb-m-%d"><rect width="%f" height="200" rx="30" fill="#FFF"/></mask>
  <g mask="url(#ocaml-ocb-m-%d)">
    <rect width="%f" height="200" fill="#%a"/>
    <rect width="%f" height="200" fill="#%a" x="%f"/>
    <rect width="%f" height="200" fill="url(#ocaml-ocb-a-%d)"/>
  </g>
  <g aria-hidden="true" fill="#fff" text-anchor="start" font-family="Verdana,DejaVu Sans,sans-serif" font-size="110">
    <text x="%f" y="148" textLength="%f" fill="#000" opacity="0.25">%s</text>
    <text x="%f" y="138" textLength="%f">%s</text>
    <text x="%f" y="148" textLength="%f" fill="#000" opacity="0.25">%s</text>
    <text x="%f" y="138" textLength="%f">%s</text>
  </g>
  %a</svg>@.|}
        xlink svg_width svg_height width accessible_text accessible_text
        !counter !counter width !counter sb_rect_width Color.pp label_color
        st_rect_width Color.pp color sb_rect_width width !counter
        (sb_text_start +. 10.) sb_text_width label sb_text_start sb_text_width
        label (sb_rect_width +. 55.) st_text_width status (sb_rect_width +. 45.)
        st_text_width status Icon.pp (icon, icon_width, 130.)
    | Flat ->
      Format.fprintf fmt
        {|<svg xmlns="http://www.w3.org/2000/svg"%s width="%f" height="%f" viewBox="0 0 %f 200" role="img" aria-label="%s">
  <title>%s</title>
  <g>
    <rect fill="#%a" width="%f" height="200"/>
    <rect fill="#%a" x="%f" width="%f" height="200"/>
  </g>
  <g aria-hidden="true" fill="#fff" text-anchor="start" font-family="Verdana,DejaVu Sans,sans-serif" font-size="110">
    <text x="%f" y="148" textLength="%f" fill="#000" opacity="0.1">%s</text>
    <text x="%f" y="138" textLength="%f">%s</text>
    <text x="%f" y="148" textLength="%f" fill="#000" opacity="0.1">%s</text>
    <text x="%f" y="138" textLength="%f">%s</text>
  </g>
  %a</svg>@.|}
        xlink svg_width svg_height width accessible_text accessible_text
        Color.pp label_color sb_rect_width Color.pp color sb_rect_width
        st_rect_width (sb_text_start +. 10.) sb_text_width label sb_text_start
        sb_text_width label (sb_rect_width +. 55.) st_text_width status
        (sb_rect_width +. 45.) st_text_width status Icon.pp
        (icon, icon_width, 132.)


src/color.ml
type t =
  | Green
  | Blue
  | Red
  | Yellow
  | Orange
  | Purple
  | Pink
  | Grey
  | Cyan
  | Black
  | Custom of string

let pp fmt = function
  | Green -> Format.fprintf fmt "3C1"
  | Blue -> Format.fprintf fmt "08C"
  | Red -> Format.fprintf fmt "E43"
  | Yellow -> Format.fprintf fmt "DB1"
  | Orange -> Format.fprintf fmt "F73"
  | Purple -> Format.fprintf fmt "94E"
  | Pink -> Format.fprintf fmt "E5B"
  | Grey -> Format.fprintf fmt "999"
  | Cyan -> Format.fprintf fmt "1BC"
  | Black -> Format.fprintf fmt "2A2A2A"
  | Custom s -> Format.fprintf fmt "%s" s

(* TODO: check s ? *)

let of_string = function
  | "green" -> Green
  | "blue" -> Blue
  | "red" -> Red
  | "yellow" -> Yellow
  | "orange" -> Orange
  | "purple" -> Purple
  | "pink" -> Pink
  | "gray" | "grey" -> Grey
  | "cyan" -> Cyan
  | "black" -> Black
  | custom -> Custom custom


src/main.ml
open Ocb

let () =
  let label = ref "label" in
  let color = ref Color.Blue in
  let style = ref Style.Classic in
  let icon = ref None in
  let icon_width = ref 13. in
  let label_color = ref (Color.Custom "555") in
  let scale = ref 1. in
  let status = ref "status" in

  let speclist =
    [ ("--label", Arg.String (fun s -> label := s), "badge label")
    ; ( "--color"
      , Arg.String (fun s -> color := Color.of_string s)
      , "badge color" )
    ; ( "--style"
      , Arg.String (fun s -> style := Style.of_string s)
      , "badge style" )
    ; ( "--labelcolor"
      , Arg.String (fun s -> label_color := Color.of_string s)
      , "label color" )
    ; ("--status", Arg.String (fun s -> status := s), "status")
    ; ("--scale", Arg.Float (fun f -> scale := f), "scale")
    ; ("--icon", Arg.String (fun s -> icon := Icon.of_string s), "icon")
    ; ("--iconwidth", Arg.Float (fun f -> icon_width := f), "icon width")
    ]
  in

  Arg.parse speclist
    (fun s -> failwith @@ Format.sprintf "unknown arg `%s`" s)
    "bla";

  let label = !label in
  let color = !color in
  let style = !style in
  let icon = !icon in
  let icon_width = !icon_width in
  let label_color = !label_color in
  let scale = !scale in
  let status = !status in

  Gen.mk Format.std_formatter ~label ~color ~style ~icon ~icon_width ~scale
    ~label_color ~status ()
